profiler/src/ProfilerEngine/Datadog.Profiler.Native/LiveObjectsProvider.cpp

```diff
@@ -0,0 +1,223 @@
+        pCorProfilerInfo,
+        pManagedThreadList,
+        pFrameStore,
+        pThreadsCpuManager,
+        pAppDomainStore,
+        pRuntimeIdStore,
+        pConfiguration,
+        nullptr);
+}
+
+const char* LiveObjectsProvider::GetName()
+{
+    return "LiveObjectsProvider";
+}
+
+void LiveObjectsProvider::OnGarbageCollectionStart(
+    int32_t number,
+    uint32_t generation,
+    GCReason reason,
+    GCType type
+    )
+{
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+
+    // address provided during AllocationTick event were not pointing to real objects
+    // so we have to wait for the next garbage collection to create a wrapping weak handle
+    for (auto& info : _objectsToMonitor)
+    {
+        auto handle = CreateWeakHandle(info.GetAddress());
+
+        if (handle != nullptr)
+        {
+            info.SetHandle(handle);
+        }
+        else
+        {
+            // this should never happen
+        }
+    }
+
+    _monitoredObjects.splice(_monitoredObjects.end(), _objectsToMonitor);
+}
+
+void LiveObjectsProvider::OnGarbageCollectionEnd(
+    int32_t number,
+    uint32_t generation,
+    GCReason reason,
+    GCType type,
+    bool isCompacting,
+    uint64_t pauseDuration,
+    uint64_t totalDuration,
+    uint64_t endTimestamp
+    )
+{
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+
+    // it is now time to check if the monitored allocated objects have been collected or are still alive
+    _monitoredObjects.remove_if([this](LiveObjectInfo& info) {
+        bool hasBeenCollected = !IsAlive(info.GetHandle());
+        if (hasBeenCollected)
+        {
+            CloseWeakHandle(info.GetHandle());
+        }
+        else
+        {
+            info.IncrementGC();
+        }
+        return hasBeenCollected;
+    });
+}
+
+std::list<std::shared_ptr<Sample>> LiveObjectsProvider::GetSamples()
+{
+    // return the live object samples
+    std::list<std::shared_ptr<Sample>> liveObjectsSamples;
+
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+    for (auto const& info : _monitoredObjects)
+    {
+        // only gen2 objects are candidates for leaking: gen0 and gen1 could be transient
+        if (info.IsGen2())
+        {
+            liveObjectsSamples.push_back(info.GetSample());
+        }
+    }
+
+    // update samples lifetime
+    int64_t currentTimestamp = OpSysTools::GetHighPrecisionTimestamp();
+    for (auto& sample : liveObjectsSamples)
+    {
+        sample->ReplaceLastLabel(Label{Sample::ObjectLifetimeLabel, std::to_string(sample->GetTimeStamp() - currentTimestamp)});
+    }
+
+    return liveObjectsSamples;
+}
+
+void LiveObjectsProvider::OnAllocation(RawAllocationSample& rawSample)
+{
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+
+    LiveObjectInfo info(
+        _pAllocationsProvider->TransformRawSample(rawSample),
+        rawSample.Address,
+        rawSample.Timestamp);
```

Comment:
You can do the creation of `LiveObjectInfo` objects, if there is enough space in the queue and if the handle is not nullptr.

Reply:
✅ Good point